File: sdk/python/src/agentfs/toolcalls.py
"""Tool call tracking and statistics."""
import json
from dataclasses import dataclass
from typing import Any, Optional, List, Literal, TYPE_CHECKING

if TYPE_CHECKING:
    from .turso_async import AsyncTursoConnection


@dataclass
class ToolCall:
    """Recorded tool call."""
    id: int
    name: str
    parameters: Optional[Any]
    result: Optional[Any]
    error: Optional[str]
    status: Literal['pending', 'success', 'error']
    started_at: int
    completed_at: Optional[int]
    duration_ms: Optional[int]


@dataclass
class ToolCallStats:
    """Aggregated statistics for a tool."""
    name: str
    total_calls: int
    successful: int
    failed: int
    avg_duration_ms: float


class ToolCalls:
    """Async tool call tracking."""

    def __init__(self, db: "AsyncTursoConnection"):
        self._db = db
        self._ready = False

    async def ready(self) -> None:
        """Initialize database schema."""
        if self._ready:
            return

        await self._db.execute("""
            CREATE TABLE IF NOT EXISTS tool_calls (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL,
                parameters TEXT,
                result TEXT,
                error TEXT,
                status TEXT NOT NULL DEFAULT 'pending',
                started_at INTEGER NOT NULL,
                completed_at INTEGER,
                duration_ms INTEGER
            )
        """)
        await self._db.execute("""
            CREATE INDEX IF NOT EXISTS idx_tool_calls_name
            ON tool_calls(name)
        """)
        await self._db.execute("""
            CREATE INDEX IF NOT EXISTS idx_tool_calls_started_at
            ON tool_calls(started_at)
        """)
        await self._db.commit()
        self._ready = True

    async def start(self, name: str, parameters: Optional[Any] = None) -> int:
        """Start tracking a tool call."""
        import time
        params_json = json.dumps(parameters) if parameters is not None else None
        cursor = await self._db.execute("""
            INSERT INTO tool_calls (name, parameters, started_at, status)
            VALUES (?, ?, ?, 'pending')
        """, (name, params_json, int(time.time())))
        await self._db.commit()
        return cursor.lastrowid

    async def success(self, call_id: int, result: Optional[Any] = None) -> None:
        """Mark tool call as successful."""
        import time
        result_json = json.dumps(result) if result is not None else None
        completed_at = int(time.time())

        # Get started_at to calculate duration
        cursor = await self._db.execute("""
            SELECT started_at FROM tool_calls WHERE id = ?
        """, (call_id,))
        row = await cursor.fetchone()
        if row:
            duration_ms = (completed_at - row[0]) * 1000
        else:
            duration_ms = None

        await self._db.execute("""
            UPDATE tool_calls
            SET status = 'success', result = ?, completed_at = ?, duration_ms = ?
            WHERE id = ?
        """, (result_json, completed_at, duration_ms, call_id))
        await self._db.commit()

    async def error(self, call_id: int, error: str) -> None:
        """Mark tool call as failed."""
        import time
        completed_at = int(time.time())

        # Get started_at to calculate duration
        cursor = await self._db.execute("""
            SELECT started_at FROM tool_calls WHERE id = ?
        """, (call_id,))
        row = await cursor.fetchone()
        if row:
            duration_ms = (completed_at - row[0]) * 1000
        else:
            duration_ms = None

        await self._db.execute("""
            UPDATE tool_calls
            SET status = 'error', error = ?, completed_at = ?, duration_ms = ?
            WHERE id = ?
        """, (error, completed_at, duration_ms, call_id))
        await self._db.commit()

    async def record(
        self,
        name: str,
        started_at: int,
        completed_at: int,
        parameters: Optional[Any] = None,
        result: Optional[Any] = None,
        error: Optional[str] = None
    ) -> int:
        """Record a completed tool call."""
        params_json = json.dumps(parameters) if parameters is not None else None
        result_json = json.dumps(result) if result is not None else None
        status = 'error' if error else 'success'
        duration_ms = (completed_at - started_at) * 1000

        cursor = await self._db.execute("""
            INSERT INTO tool_calls
            (name, parameters, result, error, status, started_at, completed_at, duration_ms)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """, (name, params_json, result_json, error, status, started_at, completed_at, duration_ms))
        await self._db.commit()
        return cursor.lastrowid

    async def get(self, call_id: int) -> Optional[ToolCall]:
        """Get a specific tool call."""
        cursor = await self._db.execute("""
            SELECT id, name, parameters, result, error, status,
                   started_at, completed_at, duration_ms
            FROM tool_calls WHERE id = ?
        """, (call_id,))
        row = await cursor.fetchone()
        if row is None:
            return None
        return self._row_to_toolcall(row)

    async def get_by_name(self, name: str, limit: int = 100) -> List[ToolCall]:
        """Get tool calls by name."""
        cursor = await self._db.execute("""
            SELECT id, name, parameters, result, error, status,
                   started_at, completed_at, duration_ms
            FROM tool_calls
            WHERE name = ?
            ORDER BY started_at DESC
            LIMIT ?
        """, (name, limit))
        rows = await cursor.fetchall()
        return [self._row_to_toolcall(row) for row in rows]

    async def get_recent(self, since: int, limit: int = 100) -> List[ToolCall]:
        """Get recent tool calls."""
        cursor = await self._db.execute("""
            SELECT id, name, parameters, result, error, status,
                   started_at, completed_at, duration_ms
            FROM tool_calls
            WHERE started_at >= ?
            ORDER BY started_at DESC
            LIMIT ?
        """, (since, limit))
        rows = await cursor.fetchall()
        return [self._row_to_toolcall(row) for row in rows]

    async def get_stats(self) -> List[ToolCallStats]:
        """Get aggregated statistics."""
        cursor = await self._db.execute("""
            SELECT
                name,
                COUNT(*) as total_calls,
                SUM(CASE WHEN status = 'success' THEN 1 ELSE 0 END) as successful,
                SUM(CASE WHEN status = 'error' THEN 1 ELSE 0 END) as failed,
                AVG(CASE WHEN status != 'pending' THEN duration_ms ELSE NULL END) as avg_duration_ms
            FROM tool_calls
            GROUP BY name
        """)
        rows = await cursor.fetchall()
        return [
            ToolCallStats(
                name=row[0],
                total_calls=row[1],
                successful=row[2],
                failed=row[3],
                avg_duration_ms=row[4] or 0.0
            )
            for row in rows
        ]

    def _row_to_toolcall(self, row) -> ToolCall:
        """Convert database row to ToolCall."""
        return ToolCall(
            id=row[0],
            name=row[1],
            parameters=json.loads(row[2]) if row[2] else None,
            result=json.loads(row[3]) if row[3] else None,
            error=row[4],
            status=row[5],
            started_at=row[6],
            completed_at=row[7],
            duration_ms=row[8]
        )

File: sdk/go/toolcalls.go
package agentfs

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	_ "turso.tech/database/tursogo"
)

type ToolCallStatus string

const (
	Pending ToolCallStatus = "pending"
	Success ToolCallStatus = "success"
	Error   ToolCallStatus = "error"
)

type ToolCall struct {
	Id          int
	Name        string
	Parameters  any
	Result      any
	Error       *string
	Status      ToolCallStatus
	StartedAt   *int
	CompletedAt *int
	DurationMs  *int
}

type ToolCallStats struct {
	Name          string
	TotalCalls    int
	Successful    int
	Failed        int
	AvgDurationMs float64
}

type ToolCalls struct {
	db *sql.DB
}

func NewToolCalls(db *sql.DB) *ToolCalls {
	return &ToolCalls{
		db: db,
	}
}

func NewToolCallsFromDatabase(db *sql.DB) (*ToolCalls, error) {
	toolCalls := NewToolCalls(db)
	err := toolCalls.Initialize()
	if err != nil {
		return nil, err
	}
	return toolCalls, nil
}

func (tc *ToolCalls) Initialize() (err error) {
	ctx := context.Background()
	_, err = tc.db.ExecContext(ctx, `
      CREATE TABLE IF NOT EXISTS tool_calls (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        name TEXT NOT NULL,
        parameters TEXT,
        result TEXT,
        error TEXT,
        status TEXT NOT NULL DEFAULT 'pending',
        started_at INTEGER NOT NULL,
        completed_at INTEGER,
        duration_ms INTEGER
      )
    `)
	if err != nil {
		return
	}
	_, err = tc.db.ExecContext(ctx, `
      CREATE INDEX IF NOT EXISTS idx_tool_calls_name
      ON tool_calls(name)
    `)
	if err != nil {
		return
	}

	_, err = tc.db.ExecContext(ctx, `
      CREATE INDEX IF NOT EXISTS idx_tool_calls_started_at
      ON tool_calls(started_at)
    `)
	return
}

func (tc *ToolCalls) Start(name string, parameters any) (id int, err error) {
	var serializedParameters *string
	switch parameters {
	case nil:
		serializedParameters = nil
	default:
		serParams, err := json.Marshal(parameters)
		if err != nil {
			return 0, err
		}
		serParamsStr := string(serParams)
		serializedParameters = &serParamsStr
	}
	startedAt := int64(time.Now().UnixMilli() / 1000)
	ctx := context.Background()
	stmt, err := tc.db.PrepareContext(ctx, `
      INSERT INTO tool_calls (name, parameters, status, started_at)
      VALUES (?, ?, 'pending', ?)
      RETURNING id
    `)
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	row := stmt.QueryRowContext(ctx, name, serializedParameters, startedAt)
	err = row.Scan(&id)
	return
}

func (tc *ToolCalls) Success(id int, result any) (err error) {
	var serializedRes *string
	switch result {
	case nil:
		serializedRes = nil
	default:
		serRes, err := json.Marshal(result)
		if err != nil {
			return err
		}
		serResStr := string(serRes)
		serializedRes = &serResStr
	}
	completedAt := int64(time.Now().UnixMilli() / 1000)
	ctx := context.Background()
	stmt, err := tc.db.PrepareContext(ctx, `SELECT started_at FROM tool_calls WHERE id = ?`)
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	row := stmt.QueryRowContext(ctx, id)
	var startedAt int64
	err = row.Scan(&startedAt)
	if err != nil {
		return
	}
	durationMs := (completedAt - startedAt) * 1000
	updateStmt, err := tc.db.PrepareContext(ctx, `
      UPDATE tool_calls
      SET status = 'success', result = ?, completed_at = ?, duration_ms = ?
      WHERE id = ?
    `)
	if err != nil {
		return err
	}
	defer func() { err = updateStmt.Close() }()
	_, err = updateStmt.ExecContext(ctx, serializedRes, completedAt, durationMs, id)
	return
}

func (tc *ToolCalls) Error(id int, errorMsg string) (err error) {
	completedAt := int64(time.Now().UnixMilli() / 1000)
	ctx := context.Background()
	stmt, err := tc.db.PrepareContext(ctx, `SELECT started_at FROM tool_calls WHERE id = ?`)
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	row := stmt.QueryRowContext(ctx, id)
	var startedAt int64
	err = row.Scan(&startedAt)
	if err != nil {
		return
	}
	durationMs := (completedAt - startedAt) * 1000
	updateStmt, err := tc.db.PrepareContext(ctx, `
      UPDATE tool_calls
      SET status = 'error', error = ?, completed_at = ?, duration_ms = ?
      WHERE id = ?
    `)
	if err != nil {
		return err
	}
	defer func() { err = updateStmt.Close() }()
	_, err = updateStmt.ExecContext(ctx, errorMsg, completedAt, durationMs, id)
	return
}

func (tc *ToolCalls) Record(
	name string,
	startedAt,
	completedAt int,
	parameters,
	result any,
	errorMsg *string,
) (id int, err error) {
	var serializedRes *string
	switch result {
	case nil:
		serializedRes = nil
	default:
		serRes, err := json.Marshal(result)
		if err != nil {
			return 0, err
		}
		serResStr := string(serRes)
		serializedRes = &serResStr
	}
	var serializedParameters *string
	switch parameters {
	case nil:
		serializedParameters = nil
	default:
		serParams, err := json.Marshal(parameters)
		if err != nil {
			return 0, err
		}
		serParamsStr := string(serParams)
		serializedParameters = &serParamsStr
	}
	durationMs := (completedAt - startedAt) * 1000
	var status ToolCallStatus
	if errorMsg == nil {
		status = Success
	} else {
		status = Error
	}
	ctx := context.Background()
	stmt, err := tc.db.PrepareContext(ctx, `
      INSERT INTO tool_calls (name, parameters, result, error, status, started_at, completed_at, duration_ms)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?)
      RETURNING id
    `)
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	row := stmt.QueryRowContext(ctx, name, serializedParameters, serializedRes, errorMsg, status, startedAt, completedAt, durationMs)
	err = row.Scan(&id)
	return
}

func (tc *ToolCalls) Get(id int) (toolCall ToolCall, err error) {
	ctx := context.Background()
	stmt, err := tc.db.PrepareContext(ctx, `
      SELECT * FROM tool_calls WHERE id = ?
    `)
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	row := stmt.QueryRowContext(ctx, id)
	err = row.Scan(
		&toolCall.Id,
		&toolCall.Name,
		&toolCall.Parameters,
		&toolCall.Error,
		&toolCall.Status,
		&toolCall.StartedAt,
		&toolCall.CompletedAt,
		&toolCall.DurationMs,
	)
	return
}

func (tc *ToolCalls) GetByName(name string, limit *int) (toolCalls []ToolCall, err error) {
	ctx := context.Background()
	var limitClause string
	switch limit {
	case nil:
	default:
		limitClause = fmt.Sprintf("LIMIT %d", *limit)
	}
	stmt, err := tc.db.PrepareContext(ctx, fmt.Sprintf(`
	SELECT * FROM tool_calls
      WHERE name = ?
      ORDER BY started_at DESC
      %s
	`, limitClause))
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	rows, err := stmt.QueryContext(ctx, name)
	if err != nil {
		return
	}
	results := []ToolCall{}
	for rows.Next() {
		var toolCall ToolCall
		err = rows.Scan(
			&toolCall.Id,
			&toolCall.Name,
			&toolCall.Parameters,
			&toolCall.Error,
			&toolCall.Status,
			&toolCall.StartedAt,
			&toolCall.CompletedAt,
			&toolCall.DurationMs,
		)
		if err != nil {
			return
		}
		results = append(results, toolCall)
	}
	toolCalls = results
	return
}

func (tc *ToolCalls) GetRecent(since int, limit *int) (toolCalls []ToolCall, err error) {
	ctx := context.Background()
	var limitClause string
	switch limit {
	case nil:
	default:
		limitClause = fmt.Sprintf("LIMIT %d", *limit)
	}
	stmt, err := tc.db.PrepareContext(ctx, fmt.Sprintf(`
	SSELECT * FROM tool_calls
      WHERE started_at > ?
      ORDER BY started_at DESC
      %s
	`, limitClause))
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	rows, err := stmt.QueryContext(ctx, since)
	if err != nil {
		return
	}
	results := []ToolCall{}
	for rows.Next() {
		var toolCall ToolCall
		err = rows.Scan(
			&toolCall.Id,
			&toolCall.Name,
			&toolCall.Parameters,
			&toolCall.Error,
			&toolCall.Status,
			&toolCall.StartedAt,
			&toolCall.CompletedAt,
			&toolCall.DurationMs,
		)
		if err != nil {
			return
		}
		results = append(results, toolCall)
	}
	toolCalls = results
	return
}

func (tc *ToolCalls) GetStats() (toolCallStats []ToolCallStats, err error) {
	ctx := context.Background()
	stmt, err := tc.db.PrepareContext(ctx, `
      SELECT
        name,
        COUNT(*) as total_calls,
        SUM(CASE WHEN status = 'success' THEN 1 ELSE 0 END) as successful,
        SUM(CASE WHEN status = 'error' THEN 1 ELSE 0 END) as failed,
        AVG(duration_ms) as avg_duration_ms
      FROM tool_calls
      WHERE status != 'pending'
      GROUP BY name
      ORDER BY total_calls DESC
    `)
	if err != nil {
		return
	}
	defer func() { err = stmt.Close() }()
	rows, err := stmt.QueryContext(ctx)
	if err != nil {
		return
	}
	results := []ToolCallStats{}
	for rows.Next() {
		var result ToolCallStats
		err = rows.Scan(
			&result.Name,
			&result.TotalCalls,
			&result.Successful,
			&result.Failed,
			&result.AvgDurationMs,
		)
		if err != nil {
			return
		}
	}
	toolCallStats = results
	return
}
